File: Project-source/terrane-server/app/api/v1/license.py
"""Frontend License area — status read-only (licensing.md: locked-state exception path; the frontend has no activation capability).

The frontend lock page polls this endpoint (locked 3s / active 8s): activation reflects near-instantly, revocation locks within ≤8s.
The activation action lives only in the admin console (terrane-admin-api).
"""

from __future__ import annotations

import asyncio
import datetime

from fastapi import APIRouter, Request

from app.licensing.state import LicenseState

router = APIRouter(prefix="/api/v1/license", tags=["license"])


def _days_left(active_until: str | None) -> int | None:
    if not active_until:
        return None
    try:
        until = datetime.datetime.fromisoformat(active_until.replace("Z", "+00:00"))
    except ValueError:
        return None
    return (until - datetime.datetime.now(datetime.timezone.utc)).days


@router.get("/status")
async def license_status(request: Request) -> dict:
    state: LicenseState = request.app.state.license
    # Throttled re-verification: locked state ≤4s (activation reflects near-instantly, but not every poll hits the edge — prevents rate-limiting/infinite loops);
    # active state ≤8s (revocation reflects near-instantly). A serial lock already prevents concurrent re-entry.
    await asyncio.to_thread(state.verify_if_stale, 4.0 if not state.unlocked else 8.0)
    verdict, payload = state.verdict, state.verdict.payload or {}
    active_until = payload.get("active_until")
    return {
        "data": {
            "required": state.required,  # false in the open-source edition → frontend hides activation/badges and guards let through
            "status": verdict.status,
            "unlocked": verdict.unlocked,
            "active_until": active_until,
            "days_left": _days_left(active_until),
        },
        "request_id": request.state.request_id,
    }

File: Project-source/terrane-web/src/lib/license.ts
/** Frontend License status, read-only (03-api.md: locked-state exception route /api/v1/license/status).
 *  The frontend has no activation rights — it only polls status; activation happens in the admin console. */
import { request } from "@/lib/api";

export interface LicenseStatus {
  status: "active" | "expiring" | "expired" | "revoked" | "binding_mismatch" | "invalid_signature" | "locked";
  /** Whether gating is enabled: the open-source backend returns false → frontend is fully unlocked
   *  (no lock page / no activation badge).
   *  When absent (old backend / undefined while loading), treated as commercial mode, preserving the original guard behavior. */
  required?: boolean;
  unlocked: boolean;
  active_until: string | null;
  days_left: number | null;
}

interface Envelope {
  data: LicenseStatus;
  request_id: string;
}

export function getLicenseStatus(): Promise<LicenseStatus> {
  return request<Envelope>("/api/v1/license/status").then((r) => r.data);
}
